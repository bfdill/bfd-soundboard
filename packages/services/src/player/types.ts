import { ChildProcess } from 'child_process'

export type DefaultPlayer =
  | 'mplayer'
  | 'afplay'
  | 'mpg123'
  | 'mpg321'
  | 'play'
  | 'omxplayer'
  | 'aplay'
  | 'cmdmp3'
export type What = string | string[]
export type PlayerOptions = {
  player: any
  players?: DefaultPlayer | DefaultPlayer[] | string[]
}
export type PlayerArgs = { [playerName: string]: string[] }
export type PlayOptions = PlayerArgs & { stdio: 'ignore' }
export type Next = Function
export interface IPlayer {
  play: (what: What, options?: PlayOptions | Next, next?: Next) => ChildProcess
}
export type createPlayer = (opts?: PlayerOptions) => IPlayer
