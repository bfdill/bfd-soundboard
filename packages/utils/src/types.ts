export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : T[P] extends (string | number | boolean)
    ? T[P]
    : DeepPartial<T[P]>
}
