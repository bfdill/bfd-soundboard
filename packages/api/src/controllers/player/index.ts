import * as koaRouter from 'koa-router'
import { NOT_FOUND, BAD_REQUEST, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '../../modules/sounds'
import { IPlayerService, PlayerService } from '../../modules/player'
import { randomIntFromInterval } from '@bfd/utils'
import { ApplicationState, IApplicationContext } from '../../types'

export interface IPlayerController {
  playSound: (context: IApplicationContext) => PromiseLike<void>
  playRando: (context: IApplicationContext) => PromiseLike<void>
}

export class PlayerController implements IPlayerController {
  private static instance: IPlayerController | undefined

  constructor(
    protected readonly playerService: IPlayerService,
    protected readonly soundService: ISoundService
  ) {}

  playSound = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playSound'
      }
    })
    const { soundId } = context.params
    logger.info({ soundId })

    if (typeof soundId !== 'string') {
      logger.error({
        soundId,
        message: 'soundId !== string'
      })
      context.body = { soundId }
      context.status = BAD_REQUEST
      return
    }

    const sound = this.soundService.getBySoundId(soundId, logger)

    if (sound === undefined) {
      logger.error({
        soundId,
        message: 'sound === undefined'
      })
      context.body = { soundId }
      context.status = NOT_FOUND
      return
    }

    logger.debug({
      sound,
      soundId,
      message: 'playFile'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  playRando = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playRando'
      }
    })
    logger.info('request')
    const sounds = this.soundService.getSounds(logger)

    if (sounds.length === 0) {
      logger.error({
        message: 'sounds.length === 0'
      })
      context.status = NOT_FOUND
      return
    }

    logger.silly({
      soundLength: sounds.length,
      message: 'sounds length'
    })

    const sound = sounds[randomIntFromInterval(0, sounds.length - 1)]

    logger.debug({
      sound,
      message: 'selected sound'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  static getInstance(): IPlayerController {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerController(
      PlayerService.getInstance(),
      SoundService.getInstance()
    )

    return this.instance
  }
}

export const getPlayerRouter = () =>
  new koaRouter<ApplicationState, IApplicationContext>()
    .get('/rando', PlayerController.getInstance().playRando)
    .get('/:soundId', PlayerController.getInstance().playSound)
