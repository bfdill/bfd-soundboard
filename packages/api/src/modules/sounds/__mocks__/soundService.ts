import { ISoundService, Sound } from '..'

export interface IMockSoundService extends ISoundService {
  setGetBySoundId: (result: any) => void
  setGetSounds: (result: any) => void
  setIsPathValid: (result: boolean) => void
  getBySoundIdMock: () => jest.Mock
  getSoundsMock: (result: any) => jest.Mock
  IsPathValidMock: (result: boolean) => jest.Mock
}

export const getMockSoundService = (): IMockSoundService => {
  const getBySoundId = jest.fn()
  const getSounds = jest.fn()
  const isPathValid = jest.fn()

  return {
    getBySoundId,
    getSounds,
    isPathValid,
    getBySoundIdMock: getBySoundId,
    getSoundsMock: getSounds,
    IsPathValidMock: isPathValid,
    setGetBySoundId: (result: Sound | undefined) =>
      getBySoundId.mockReturnValue(result),
    setGetSounds: (result: Sound[]) => getSounds.mockReturnValue(result),
    setIsPathValid: (result: boolean) => isPathValid.mockReturnValue(result)
  } as any
}
