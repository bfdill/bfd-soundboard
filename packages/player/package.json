{
  "name": "@talkboy-9000/player",
  "version": "0.0.1",
  "license": "Unlicense",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -p ./tsconfig.json",
    "clean": "rimraf dist",
    "test": "jest --no-cache",
    "watch": "yarn build --watch --preserveWatchOutput"
  },
  "dependencies": {
    "@talkboy-9000/sounds": "^0.0.1"
  },
  "devDependencies": {
    "@talkboy-9000/models": "^0.0.1",
    "@talkboy-9000/winston-jest": "^0.0.1"
  }
}
