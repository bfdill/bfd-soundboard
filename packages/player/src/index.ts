export * from './types'
export * from './player.types'

export { PlayerService } from './player'
