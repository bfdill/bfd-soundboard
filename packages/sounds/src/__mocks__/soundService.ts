import { Sound } from '@talkboy-9000/models'
import { ISoundService } from '../sounds.types'

export interface IMockSoundService extends ISoundService {
  setGetBySoundId: (result: any) => void
  setGetSounds: (result: any) => void
  setIsPathValid: (result: boolean) => void
  getBySoundIdMock: () => jest.Mock
  getSoundsMock: (result: any) => jest.Mock
  IsPathValidMock: (result: boolean) => jest.Mock
}

export const getMockSoundService = (): IMockSoundService => {
  const getBySoundId = jest.fn()
  const getSounds = jest.fn()
  const isPathValid = jest.fn()

  return <IMockSoundService>{
    getBySoundId,
    getSounds,
    isPathValid,
    getBySoundIdMock: getBySoundId,
    getSoundsMock: getSounds,
    IsPathValidMock: isPathValid,
    setGetBySoundId: (result: Sound | undefined) =>
      getBySoundId.mockReturnValue(result),
    setGetSounds: (result: Sound[]) => getSounds.mockReturnValue(result),
    setIsPathValid: (result: boolean) => isPathValid.mockReturnValue(result)
  }
}
