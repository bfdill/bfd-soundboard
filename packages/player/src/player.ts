// import * as playSound from 'play-sound'
import * as winston from 'winston'
import { createPlayer, IPlayer } from './types'
import { IPlayerService } from './player.types'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

export class PlayerService implements IPlayerService {
  static instance: IPlayerService | undefined

  constructor(
    protected readonly player: IPlayer,
    protected readonly soundService: ISoundService
  ) {}

  playFile = (
    filename: string,
    parentLogger: winston.Logger
  ): PromiseLike<void> => {
    const logger = parentLogger.child({
      service: {
        name: 'PlayerService',
        method: 'playFile'
      }
    })

    logger.info({ filename })

    if (!this.soundService.isPathValid(filename, logger)) {
      const message = 'filename path is invalid'
      logger.error({ filename, message })
      return Promise.reject(message)
    }

    return new Promise((resolve, reject) => {
      const next = (error?: any) => {
        if (error === undefined || error === null) {
          logger.debug({
            filename,
            message: 'playback success'
          })
          resolve()
          return
        }

        logger.debug({
          error,
          filename,
          message: 'playback failure'
        })
        reject(error)
      }

      this.player.play(filename, next)
    })
  }

  static getInstance() {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerService(
      createPlayer(),
      SoundService.getInstance()
    )

    return this.instance
  }
}
