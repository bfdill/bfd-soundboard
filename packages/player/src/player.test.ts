import { ChildProcess } from 'child_process'
import { getJestLogger, IJestLogger } from '@talkboy-9000/winston-jest'
import { getMockSoundService } from '../../sounds/src/__mocks__/soundService'

import { createPlayer, IPlayer, What, PlayOptions, Next } from './types'
import { IPlayerService } from './player.types'
import { PlayerService } from './player'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

describe('playerService', () => {
  const filename = 'this-file-totes-exists.mp3'
  const mockNextInput: jest.Mock = jest.fn()
  const fakePlay = (
    _0: What,
    options?: PlayOptions | Next,
    _2?: Next
  ): ChildProcess => {
    if (typeof options === 'function') {
      options(mockNextInput())
    }

    return jest.fn() as any
  }
  const player: IPlayer = {
    play: fakePlay
  }
  const mockSoundService = getMockSoundService()
  const jestLogger: IJestLogger = getJestLogger()
  const playerService: IPlayerService = new PlayerService(
    player,
    mockSoundService
  )

  beforeEach(() => {
    mockSoundService.setIsPathValid(true)
    mockNextInput.mockReturnValue(null)
  })

  test('rejects on invalid file', async () => {
    expect.assertions(3)
    mockSoundService.setIsPathValid(false)

    await expect(
      playerService.playFile(filename, jestLogger.logger)
    ).rejects.toMatchSnapshot()

    jestLogger.callsMatchSnapshot()
  })

  test('rejects on playback error', async () => {
    expect.assertions(3)
    mockNextInput.mockReturnValue({ error: 'KHAAAAAAN!' })

    await expect(
      playerService.playFile(filename, jestLogger.logger)
    ).rejects.toMatchSnapshot()

    jestLogger.callsMatchSnapshot()
  })

  test('playback happy path', async () => {
    expect.assertions(3)

    await expect(
      playerService.playFile(filename, jestLogger.logger)
    ).resolves.toMatchSnapshot()

    jestLogger.callsMatchSnapshot()
  })
})
