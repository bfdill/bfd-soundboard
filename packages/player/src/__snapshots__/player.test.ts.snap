// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playerService playback happy path 1`] = `undefined`;

exports[`playerService playback happy path 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
    },
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\"},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`playerService playback happy path 3`] = `
Array [
  Object {
    "filename": "this-file-totes-exists.mp3",
    "level": "debug",
    "message": "playback success",
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"message\\":\\"playback success\\",\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`playerService rejects on invalid file 1`] = `"filename path is invalid"`;

exports[`playerService rejects on invalid file 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
    },
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\"},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`playerService rejects on invalid file 3`] = `
Array [
  Object {
    "filename": "this-file-totes-exists.mp3",
    "level": "error",
    "message": "filename path is invalid",
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "error",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"message\\":\\"filename path is invalid\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`playerService rejects on playback error 1`] = `
Object {
  "error": "KHAAAAAAN!",
}
`;

exports[`playerService rejects on playback error 2`] = `
Array [
  Object {
    "level": "info",
    "message": Object {
      "filename": "this-file-totes-exists.mp3",
    },
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "info",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"message\\":{\\"filename\\":\\"this-file-totes-exists.mp3\\"},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`playerService rejects on playback error 3`] = `
Array [
  Object {
    "error": Object {
      "error": "KHAAAAAAN!",
    },
    "filename": "this-file-totes-exists.mp3",
    "level": "debug",
    "message": "playback failure",
    "service": Object {
      "method": "playFile",
      "name": "PlayerService",
    },
    Symbol(level): "debug",
    Symbol(message): "{\\"service\\":{\\"name\\":\\"PlayerService\\",\\"method\\":\\"playFile\\"},\\"error\\":{\\"error\\":\\"KHAAAAAAN!\\"},\\"filename\\":\\"this-file-totes-exists.mp3\\",\\"message\\":\\"playback failure\\",\\"level\\":\\"debug\\"}",
  },
]
`;
